lib/Semantic_errors.ml

```diff
@@ -0,0 +1,226 @@
+open Core_kernel
+open Pretty_printing
+open Mir
+
+(** Low-level semantic errors used for validation *)
+type t =
+  | IdentifierIsKeyword of location_span * string
+  | IdentifierIsModelName of location_span * string
```

Comment:
This is a super welcome refactor! I should've done it that way the first time around, obviously. Thank you!

Reply:
I think we should consider rewriting the whole of the semantic check using these low level errors in combination with the custom validation module I've added (see my comment below on why I'm using a custom approach).

Reply:
I love having a list of all of the error types in one place! Amazing. 

Agreed about your refactor assuming Matthijs also agrees.